File: src/DomainF/Tournaments/State.fs
namespace Orleans.Tournament.Domain.Tournaments
open System

type TournamentState =
    val mutable Id:Guid
    val mutable Created:bool
    val mutable Name:string
    val mutable Teams:list<Guid>
    val mutable Fixture:Fixture option
    new() = {
        Id = Guid.Empty;
        Created = false;
        Name = String.Empty;
        Teams = []
        Fixture = None
    }
    member x.Apply (evt : TournamentCreated) =
        x.Created <- true;
        x.Id <- evt.TournamentId;
        x.Name <- evt.Name;
    member x.Apply (evt : TeamAdded) =
        x.Teams = evt.TeamId :: x.Teams
    member x.Apply (evt : TournamentStarted) =
        x.Fixture <- Some (Fixture.Create evt.Teams);
    member x.Apply (evt : MatchResultSet) =
        x.Fixture <- Some (x.Fixture.Value.SetMatchResult evt.MatchInfo);

File: src/Domain/Tournaments/State.cs
namespace Tournament.Domain.Tournaments;

public partial class TournamentState
{
    public Guid Id { get; set; }
    public bool Created { get; set; }
    public string Name { get; set; }
    public List<Guid> Teams { get; set; }
    public Fixture? Fixture { get; set; }

    public TournamentState()
    {
        Name = string.Empty;
        Teams = Enumerable.Empty<Guid>().ToList();
    }

    public void Apply(TournamentCreated evt)
    {
        Id = evt.TournamentId;
        Created = true;
    }

    public void Apply(TeamAdded evt)
        => Teams.Add(evt.TeamId);

    public void Apply(TournamentStarted evt)
        => Fixture = Fixture.Create(Teams, evt.Seed);

    public void Apply(MatchResultSet evt)
        => Fixture = Fixture!.SetMatchResult(evt.Match);
}